src/library.cpp

```diff
@@ -157,6 +157,123 @@ bool Library::removeBookmark(const std::string& zimId, const std::string& url)
+      }
+    }
+  }
+  int changed = 0;
+  for(auto& sourceBook:sourceBooks) {
+    changed += migrateBookmarks(sourceBook, ALLOW_DOWNGRADE);
+  }
+  return std::make_tuple(changed, invalidBookmarks);
+}
+
+void Library::pruneBookCollectionFromOlderBooks(BookIdCollection& books, const Bookmark& bookmark, MigrationMode migrationMode) const {
+  sort(books, DATE, false);
+
+  // Remove source book
+  if (migrationMode == UPGRADE_ONLY) {
+    auto foundIt = std::find(books.begin(), books.end(), bookmark.getBookId());
+    if (foundIt == books.end()) {
+      foundIt = std::find_if(books.begin(), books.end(), [&](std::string& bookId) {
+        auto& book = getBookById(bookId);
+        return book.getDate() <= bookmark.getDate();
+      });
+    }
+    if (foundIt != books.end()) {
+      // books is sorted by date, remove the source and all oldest books
+      books.erase(foundIt, books.end());
+    }
+  }
+}
+
+std::string remove_quote(std::string input) {
+  std::replace(input.begin(), input.end(), '"', ' ');
+  return input;
+}
+
+std::string Library::getBestTargetBookId(const Bookmark& bookmark, MigrationMode migrationMode) const {
+  std::lock_guard<std::recursive_mutex> lock(m_mutex);
+  // Search for a existing book with the same name
+  auto book_filter = Filter();
+  if (!bookmark.getBookName().empty()) {
+    book_filter.name(bookmark.getBookName());
+  } else {
+    // We don't have a name stored (older bookmarks)
+    // Fallback on title (All bookmarks should have one, but let's be safe against wrongly filled bookmark)
+    if (bookmark.getBookTitle().empty()) {
+        // No bookName nor bookTitle, no way to find target book.
+        return "";
+    }
+    book_filter.query("title:\"" + remove_quote(bookmark.getBookTitle()) + "\"");
+  }
+  auto targetBooks = filter(book_filter);
+  pruneBookCollectionFromOlderBooks(targetBooks, bookmark, migrationMode);
+    
+  if (targetBooks.empty()) {
+    // No better book to migrate to.
+    // If bookmark is valid, best is the current one.
+    try {
+      auto book = getBookById(bookmark.getBookId());
+      return bookmark.getBookId();
+    } catch (std::out_of_range& e) {
+      return "";
+    }
+  }
+  if (targetBooks.size() != 1) {
+    // We have several, reduce to same flavour
+    auto flavouredTargetBooks = filter(book_filter.flavour(bookmark.getBookFlavour()));
+    pruneBookCollectionFromOlderBooks(flavouredTargetBooks, bookmark, migrationMode);
+    if (!flavouredTargetBooks.empty()) {
+      targetBooks = flavouredTargetBooks;
+    }
+  }
+  return targetBooks[0];
+}
+
+int Library::migrateBookmarks(const std::string& sourceBookId, MigrationMode migrationMode) {
+  std::lock_guard<std::recursive_mutex> lock(m_mutex);
+
+  Bookmark firstBookmarkToChange;
+  for(auto& bookmark:m_bookmarks) {
+    if (bookmark.getBookId() == sourceBookId) {
+      firstBookmarkToChange = bookmark;
+      break;
+    }
+  }
+
+  if (firstBookmarkToChange.getBookId().empty()) {
+    return 0;
```

Comment:
Drawing attention to Codecov's discovery that "Added line #L254 was not covered by tests"